### PR TITLE
Handle array bounds of VB in IDE0032

### LINKED ISSUE #38218: VB Use Auto Property brakes code with explicit Array bounds
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/517543/vb-use-auto-property-brakes-code-with-explicit-arr.html)._

---
```
Private ReadOnly _Condition(255) As Integer
```
Message	IDE0032	Use auto property
The fix produces code with a 0 length array possibily of arrays that is not readonly

```
 Public Property Condition() As Integer()
```

---
#### Visual Studio Feedback System on 4/4/2019, 02:40 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by huoyaoyuan**
Fixes #38218

### CODE REVIEW

File: src/Features/VisualBasic/Portable/UseAutoProperty/VisualBasicUseAutoPropertyCodeFixProvider.vb [hunk 1]
```diff
@@ -63,19 +63,30 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.UseAutoProperty
                 statement = statement.WithAsClause(initializer.asNewClause)
             End If
 
+            If initializer.arrayBounds IsNot Nothing Then
+                Dim arrayCreation = SyntaxFactory.ArrayCreationExpression(
+                    Nothing,
+                    DirectCast(statement.GetReturnType(), ArrayTypeSyntax).ElementType,
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/UseAutoProperty/VisualBasicUseAutoPropertyCodeFixProvider.vb:69]: how do you that it's safe to do `DirectCast(statement.GetReturnType(), ArrayTypeSyntax)`?

Comment (CyrusNajmabadi) [reply src/Features/VisualBasic/Portable/UseAutoProperty/VisualBasicUseAutoPropertyCodeFixProvider.vb:69]: would far prefer you use semantics to get this type, and then do `.GenerateTypeSyntax()` from that.

File: src/Features/VisualBasic/Portable/UseAutoProperty/VisualBasicUseAutoPropertyCodeFixProvider.vb [hunk 2]
```diff
@@ -63,19 +63,30 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.UseAutoProperty
...
+                    initializer.arrayBounds,
+                    If(TryCast(initializer.equalsValue?.Value, CollectionInitializerSyntax), SyntaxFactory.CollectionInitializer()))
+                statement = statement.WithTrailingTrivia(SyntaxFactory.Space) _
+                    .WithInitializer(SyntaxFactory.EqualsValue(arrayCreation))
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/UseAutoProperty/VisualBasicUseAutoPropertyCodeFixProvider.vb:73]:
```suggestion
                statement = statement.WithTrailingTrivia(SyntaxFactory.Space).
                                      WithInitializer(SyntaxFactory.EqualsValue(arrayCreation))
```

Comment (huoyaoyuan) [reply src/Features/VisualBasic/Portable/UseAutoProperty/VisualBasicUseAutoPropertyCodeFixProvider.vb:73]: I'm following the style with several lines above here.

File: src/Features/VisualBasic/Portable/UseAutoProperty/VisualBasicUseAutoPropertyCodeFixProvider.vb [hunk 3]
```diff
@@ -63,19 +63,32 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.UseAutoProperty
                 statement = statement.WithAsClause(initializer.asNewClause)
             End If
 
+            If initializer.arrayBounds IsNot Nothing Then
+                Dim semanticsModel = Await propertyDocument.GetSemanticModelAsync(cancellationToken).ConfigureAwait(False)
+                Dim arrayType = DirectCast(semanticsModel.GetDeclaredSymbol(propertyDeclaration, cancellationToken).Type, IArrayTypeSymbol)
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/UseAutoProperty/VisualBasicUseAutoPropertyCodeFixProvider.vb:68]: i don't know why this cast is necessarily safe.  can you make a try/cast?

Comment (huoyaoyuan) [reply src/Features/VisualBasic/Portable/UseAutoProperty/VisualBasicUseAutoPropertyCodeFixProvider.vb:68]: Seeing `ArrayBounds` should means it's an array type. But it's OK to be defensive and robust in analyzers layer.

File: src/Analyzers/VisualBasic/Tests/UseAutoProperty/UseAutoPropertyTests.vb [hunk 4]
```diff
@@ -1238,5 +1238,56 @@ class Class1
...
+    Public ReadOnly Property Condition As Integer() = { 1, 2, 3 }
+End Class")
+        End Function
+
+        <WorkItem(38218, "https://github.com/dotnet/roslyn/issues/38218")>
+        <Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseAutoProperty)>
+        Public Async Function TestArrayRankAndInitializerAsync() As Task
+            Await TestInRegularAndScriptAsync(
+"Class C
+    [|Private ReadOnly _Condition(10) As Integer = { 1, 2, 3 }|]
```
Comment (huoyaoyuan) [inline src/Analyzers/VisualBasic/Tests/UseAutoProperty/UseAutoPropertyTests.vb:1281]: Realized that this one is actually illegal in VB, and the fixed version is illegal too. Removing.

File: src/Features/VisualBasic/Portable/UseAutoProperty/VisualBasicUseAutoPropertyCodeFixProvider.vb [hunk 5]
```diff
@@ -63,19 +63,34 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.UseAutoProperty
...
+                Dim semanticsModel = Await propertyDocument.GetSemanticModelAsync(cancellationToken).ConfigureAwait(False)
+                Dim arrayType = TryCast(semanticsModel.GetDeclaredSymbol(propertyDeclaration, cancellationToken).Type, IArrayTypeSymbol)
+                If arrayType IsNot Nothing Then
+                    Dim arrayCreation = SyntaxFactory.ArrayCreationExpression(
+                    Nothing,
+                    arrayType.ElementType.GenerateTypeSyntax(),
+                    initializer.arrayBounds,
+                    If(TryCast(initializer.equalsValue?.Value, CollectionInitializerSyntax), SyntaxFactory.CollectionInitializer()))
+                    statement = statement.WithTrailingTrivia(SyntaxFactory.Space).
+                                      WithInitializer(SyntaxFactory.EqualsValue(arrayCreation))
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/UseAutoProperty/VisualBasicUseAutoPropertyCodeFixProvider.vb:76]: Either:

```suggestion
                    statement = statement.WithTrailingTrivia(SyntaxFactory.Space).
                                          WithInitializer(SyntaxFactory.EqualsValue(arrayCreation))
```

or

```suggestion
                    statement = statement.WithTrailingTrivia(SyntaxFactory.Space).
                        WithInitializer(SyntaxFactory.EqualsValue(arrayCreation))
```

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
thanks!

**Pull request merged**